### DISCUSSION TITLE
Opinions - How to recognize "stuck" instances

### DISCUSSION THREAD

**Discussion by mpaul31** [Azure/azure-functions-durable-extension#1770] [Q&A] [answered]
I just wanted to ask and see what everyone's approach is to tracking and recognizing orchestrations and/or entities that get into an unrecoverable "stuck" state. 

I am starting to design a workflow system having to deal with sale transactions and this area makes me a little uneasy.

Thanks!

**Comment by olitomlinson** [comment, marked as answer]
(As we discussed on Twitter) I think this comes down to understanding how long your orchestrations typically take to complete.

For example, if you have a use case where your orchestrations typically complete between 5 and 10 minutes, then you can query the for running instances that are older then 10 minutes, and begin raising Azure Monitor events/metrics to let you know that something might not be right.

However, if the typical orchestration lifetime varies wildly (e.g. you have a human-in-the-loop) then it becomes harder to define what is an acceptable amount of time monitor against.

Durable Functions is unsupervised by virtue of its design, so it really is a hard problem to solve at the user-code level. A challenge with DF is that when a control messages or activity message doesn't get consumed, the orchestration halts or gets "stuck". But this isn't a problem that is unique to DF. I get it with Service Bus every now and again on other Azure Functions. Its a systemic problem with queue based systems. _Something_ needs to observe message consumption, or lack of, and raise alerts/auto-heal appropriately. I don't know who is responsible for that _something_, but I would say it probably should fall on the side of the platform provider, because _serverless_ /PaaS.

**Comment by mpaul31** [comment]
anyone else care to share? :)

**Comment by withinboredom** [comment]
It depends on what you mean by "stuck" and how much data you are ok with losing ;)

I once shipped a change to production that, due to a bug, filled the queues to several thousand each by sending the same signals to the same hundred or so entities in an infinite loop. At least 10 thousand per queue before I noticed. At the rate the app was churning through items, actual work was delayed for several days. So, I just cleared the queues. Who knows what was lost, but it was better than waiting for several days to consume all the items. Any issues arising from that could be dealt with on a case-by-case basis.

I suggest putting together some monitoring that looks at queue depth. If it gets above a certain number (depending on your functions) your customers/users will begin experiencing degradation in your service and you might have a fire on your hands. If you can't lose any data, it might be ideal to bump up the number of `partitionCount` to churn through more items faster (though I haven't tried this).

**Comment by mpaul31** [comment]
> It depends on what you mean by "stuck" and how much data you are ok with losing ;)
> 
> I once shipped a change to production that, due to a bug, filled the queues to several thousand each by sending the same signals to the same hundred or so entities in an infinite loop. At least 10 thousand per queue before I noticed. At the rate the app was churning through items, actual work was delayed for several days. So, I just cleared the queues. Who knows what was lost, but it was better than waiting for several days to consume all the items. Any issues arising from that could be dealt with on a case-by-case basis.
> 
> I suggest putting together some monitoring that looks at queue depth. If it gets above a certain number (depending on your functions) your customers/users will begin experiencing degradation in your service and you might have a fire on your hands. If you can't lose any data, it might be ideal to bump up the number of `partitionCount` to churn through more items faster (though I haven't tried this).

data loss...zero :)

@olitomlinson i actually was not even thinking about the case where messages would not get consumed (didn't realize this could happen). I was more so focusing on the orchestration instance status staying "running" when in fact the orchestration completed.

**Comment by withinboredom** [comment]
> I was more so focusing on the orchestration instance status staying "running" when in fact the orchestration completed.

I don't think that can happen since it would just pick back up where it was running before and complete it again. I may be wrong though, but that's what I "think" would happen.

**Comment by mpaul31** [comment]
> > I was more so focusing on the orchestration instance status staying "running" when in fact the orchestration completed.
> 
> I don't think that can happen since it would just pick back up where it was running before and complete it again. I may be wrong though, but that's what I "think" would happen.

No it's definitely happens (happened) if you check the issue log but it could have very well been fixed with the most recent versions.

**Comment by withinboredom** [comment]
I suppose it could happen if your orchestration is waiting on an external event and the external event gets lost. I handle this by setting a custom status on the orchestration after receiving an external event, then have a "watchdog" orchestration that makes sure nothing is waiting for external events for too long. Basically the pattern looks like

[![](https://mermaid.ink/img/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG4gIHBhcnRpY2lwYW50IFUgYXMgVXNlclxuICBwYXJ0aWNpcGFudCBIIGFzIEhUVFAgRnVuY3Rpb25cbiAgcGFydGljaXBhbnQgTyBhcyBPcmNoZXN0cmF0aW9uXG4gIHBhcnRpY2lwYW50IFcgYXMgV2F0Y2hkb2dcbiAgXG4gIE5vdGUgb3ZlciBVLEg6IFVuaGFwcHkgUGF0aFxuICBhY3RpdmF0ZSBVXG4gIFUtPj4rSDogUmVxdWVzdCB3b3JrXG4gIEgtPj4rTzogU3RhcnQgT3JjaGVzdHJhdGlvblxuICBPLT4-K1c6IFN0YXJ0IFdhdGNoZG9nXG4gIEgtLT4-VTogMnh4IHJlc3BvbnNlXG4gIGRlYWN0aXZhdGUgSFxuICBkZWFjdGl2YXRlIFVcbiAgTm90ZSBvdmVyIE86IERvIFdvcmssIHNldCBjdXN0b20gc3RhdHVzIHdhaXRpbmcgZm9yIG1lc3NhZ2VcbiAgYWN0aXZhdGUgVVxuICBhY3RpdmF0ZSBIXG4gIFUtPj4rSDogU2VuZCBNZXNzYWdlXG4gIEgteE86IFNlbmQgTWVzc2FnZVxuICBOb3RlIG92ZXIgTzogTmV2ZXIgcmVjZWl2ZSBtZXNzYWdlXG4gIEgtLT4-VTogU2VudCBNZXNzYWdlXG4gIGRlYWN0aXZhdGUgVVxuICBkZWFjdGl2YXRlIEhcbiAgVy0-TzogQ2hlY2sgc3RhdHVzXG4gIE5vdGUgb3ZlciBPLFc6IEl0J3MgYmVlbiB0b28gbG9uZ1xuICBXLT4-TzogRmFpbCBPcmNoZXN0cmF0aW9uXG4gIGRlYWN0aXZhdGUgT1xuICBkZWFjdGl2YXRlIFdcbiAgYWN0aXZhdGUgVVxuICBhY3RpdmF0ZSBIXG4gIFUtPj5IOiBDaGVjayBzdGF0dXNcbiAgSC0tPj5VOiBGYWlsZWQgT3JjaGVzdHJhdGlvblxuICBkZWFjdGl2YXRlIFVcbiAgZGVhY3RpdmF0ZSBIXG4gIE5vdGUgb3ZlciBVLEg6IEhhbmRsZSBmYWlsZWQgT3JjaGVzdHJhdGlvbiBhcyBub3JtYWwiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG4gIHBhcnRpY2lwYW50IFUgYXMgVXNlclxuICBwYXJ0aWNpcGFudCBIIGFzIEhUVFAgRnVuY3Rpb25cbiAgcGFydGljaXBhbnQgTyBhcyBPcmNoZXN0cmF0aW9uXG4gIHBhcnRpY2lwYW50IFcgYXMgV2F0Y2hkb2dcbiAgXG4gIE5vdGUgb3ZlciBVLEg6IFVuaGFwcHkgUGF0aFxuICBhY3RpdmF0ZSBVXG4gIFUtPj4rSDogUmVxdWVzdCB3b3JrXG4gIEgtPj4rTzogU3RhcnQgT3JjaGVzdHJhdGlvblxuICBPLT4-K1c6IFN0YXJ0IFdhdGNoZG9nXG4gIEgtLT4-VTogMnh4IHJlc3BvbnNlXG4gIGRlYWN0aXZhdGUgSFxuICBkZWFjdGl2YXRlIFVcbiAgTm90ZSBvdmVyIE86IERvIFdvcmssIHNldCBjdXN0b20gc3RhdHVzIHdhaXRpbmcgZm9yIG1lc3NhZ2VcbiAgYWN0aXZhdGUgVVxuICBhY3RpdmF0ZSBIXG4gIFUtPj4rSDogU2VuZCBNZXNzYWdlXG4gIEgteE86IFNlbmQgTWVzc2FnZVxuICBOb3RlIG92ZXIgTzogTmV2ZXIgcmVjZWl2ZSBtZXNzYWdlXG4gIEgtLT4-VTogU2VudCBNZXNzYWdlXG4gIGRlYWN0aXZhdGUgVVxuICBkZWFjdGl2YXRlIEhcbiAgVy0-TzogQ2hlY2sgc3RhdHVzXG4gIE5vdGUgb3ZlciBPLFc6IEl0J3MgYmVlbiB0b28gbG9uZ1xuICBXLT4-TzogRmFpbCBPcmNoZXN0cmF0aW9uXG4gIGRlYWN0aXZhdGUgT1xuICBkZWFjdGl2YXRlIFdcbiAgYWN0aXZhdGUgVVxuICBhY3RpdmF0ZSBIXG4gIFUtPj5IOiBDaGVjayBzdGF0dXNcbiAgSC0tPj5VOiBGYWlsZWQgT3JjaGVzdHJhdGlvblxuICBkZWFjdGl2YXRlIFVcbiAgZGVhY3RpdmF0ZSBIXG4gIE5vdGUgb3ZlciBVLEg6IEhhbmRsZSBmYWlsZWQgT3JjaGVzdHJhdGlvbiBhcyBub3JtYWwiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)

[![](https://mermaid.ink/img/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG4gIHBhcnRpY2lwYW50IFUgYXMgVXNlclxuICBwYXJ0aWNpcGFudCBIIGFzIEhUVFAgRnVuY3Rpb25cbiAgcGFydGljaXBhbnQgTyBhcyBPcmNoZXN0cmF0aW9uXG4gIHBhcnRpY2lwYW50IFcgYXMgV2F0Y2hkb2dcbiAgTm90ZSBvdmVyIFUsSDogSGFwcHkgUGF0aFxuICBhY3RpdmF0ZSBVXG4gIGFjdGl2YXRlIEhcbiAgVS0-Pkg6IFJlcXVlc3QgV29ya1xuICBhY3RpdmF0ZSBPXG4gIEgtPj5POiBTdGFydCBPcmNoZXN0cmF0aW9uXG4gIGFjdGl2YXRlIFdcbiAgTy0-Plc6IFN0YXJ0IFdhdGNoZG9nXG4gIEgtLT4-VTogMnh4IHJlc3BvbnNlXG4gIGRlYWN0aXZhdGUgSFxuICBkZWFjdGl2YXRlIFVcbiAgbm90ZSBvdmVyIE86IERvIHdvcmssIHNldCBjdXN0b20gc3RhdHVzIHdhaXRpbmcgZm9yIG1lc3NhZ2VcbiAgYWN0aXZhdGUgVVxuICBhY3RpdmF0ZSBIXG4gIFUtPj5IOiBTZW5kIE1lc3NhZ2VcbiAgSC0-Pk86IFNlbmQgTWVzc2FnZVxuICBub3RlIG92ZXIgTzogU2V0IGN1c3RvbSBzdGF0dXMgcmVjZWl2ZWQgbWVzc2FnZVxuICBILS0-PlU6IFNlbnQgTWVzc2FnZVxuICBkZWFjdGl2YXRlIFVcbiAgZGVhY3RpdmF0ZSBIXG4gIFctPk86IENoZWNrIFN0YXR1c1xuICBPLT5XOiBPcmNoZXN0cmF0aW9uIHN1Y2NlZWVkZWRcbiAgZGVhY3RpdmF0ZSBXXG4gIGRlYWN0aXZhdGUgT1xuICBhY3RpdmF0ZSBVXG4gIGFjdGl2YXRlIEhcbiAgVS0-Pkg6IENoZWNrIFN0YXR1c1xuICBILS0-PlU6IFN1Y2NlZWVkZWRcbiAgZGVhY3RpdmF0ZSBVXG4gIGRlYWN0aXZhdGUgSCIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In0sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG4gIHBhcnRpY2lwYW50IFUgYXMgVXNlclxuICBwYXJ0aWNpcGFudCBIIGFzIEhUVFAgRnVuY3Rpb25cbiAgcGFydGljaXBhbnQgTyBhcyBPcmNoZXN0cmF0aW9uXG4gIHBhcnRpY2lwYW50IFcgYXMgV2F0Y2hkb2dcbiAgTm90ZSBvdmVyIFUsSDogSGFwcHkgUGF0aFxuICBhY3RpdmF0ZSBVXG4gIGFjdGl2YXRlIEhcbiAgVS0-Pkg6IFJlcXVlc3QgV29ya1xuICBhY3RpdmF0ZSBPXG4gIEgtPj5POiBTdGFydCBPcmNoZXN0cmF0aW9uXG4gIGFjdGl2YXRlIFdcbiAgTy0-Plc6IFN0YXJ0IFdhdGNoZG9nXG4gIEgtLT4-VTogMnh4IHJlc3BvbnNlXG4gIGRlYWN0aXZhdGUgSFxuICBkZWFjdGl2YXRlIFVcbiAgbm90ZSBvdmVyIE86IERvIHdvcmssIHNldCBjdXN0b20gc3RhdHVzIHdhaXRpbmcgZm9yIG1lc3NhZ2VcbiAgYWN0aXZhdGUgVVxuICBhY3RpdmF0ZSBIXG4gIFUtPj5IOiBTZW5kIE1lc3NhZ2VcbiAgSC0-Pk86IFNlbmQgTWVzc2FnZVxuICBub3RlIG92ZXIgTzogU2V0IGN1c3RvbSBzdGF0dXMgcmVjZWl2ZWQgbWVzc2FnZVxuICBILS0-PlU6IFNlbnQgTWVzc2FnZVxuICBkZWFjdGl2YXRlIFVcbiAgZGVhY3RpdmF0ZSBIXG4gIFctPk86IENoZWNrIFN0YXR1c1xuICBPLT5XOiBPcmNoZXN0cmF0aW9uIHN1Y2NlZWVkZWRcbiAgZGVhY3RpdmF0ZSBXXG4gIGRlYWN0aXZhdGUgT1xuICBhY3RpdmF0ZSBVXG4gIGFjdGl2YXRlIEhcbiAgVS0-Pkg6IENoZWNrIFN0YXR1c1xuICBILS0-PlU6IFN1Y2NlZWVkZWRcbiAgZGVhY3RpdmF0ZSBVXG4gIGRlYWN0aXZhdGUgSCIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In0sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)

The watchdog is generalized. I just start a watchdog orchestration and pass it the current orchestration's Id, handled in a base class. Then I just call `result = this.WaitForEvent(eventName);` and then the base class handles all the logic of setting custom statuses (which are standardized).

PS. You can also set timeouts on most orchestrations, but I think that the likelihood of two orchestrations getting stuck somewhere is less than one. I have run into issues where the watchdog orchestration is in a different task/control queue than the orchestration and the orchestration's task/control queue gets bogged down. Then the watchdog sees that the orchestration isn't working (it's just behind) and will kill it. So, there's a lot to tune there. It's not perfect.

**Comment by olitomlinson** [comment]
If you don't want to use a watchdog process, you can use the overload on `WaitForExternalEvent` to provide a timeout value and then have your application react accordingly when the timeout occurs (continue waiting by rescheduling a new `WaitForExternalEvent`, or gracefully exit).

`public virtual System.Threading.Tasks.Task WaitForExternalEvent (string name, TimeSpan timeout);`

P.s. I recommend doing this anyway as a good practice. It helps weed-out and start a discussion around those "what if" scenarios that often don't get discussed until 6 months down the line when your app enters an illegal state in production at 4am...
